runtime/gc_vlhgc/SchedulingDelegate.cpp

```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+		 * by increasing or decreasing eden by 10% 
+		 */
+		if (_extensions->dnssExpectedTimeRatioMinimum._valueSpecified > hybridEdenOverhead ) {
+			/* Shrink eden a bit */
+			edenRegionChange = edenChangeMagnitude * -1;
+		} else if (_extensions->dnssExpectedTimeRatioMaximum._valueSpecified < hybridEdenOverhead) {
+			/* Expand eden a bit */
+			edenRegionChange = edenChangeMagnitude;
+		}
+
+		_edenSizeFactor += edenRegionChange;	
+	}
+}
+
+double 
+MM_SchedulingDelegate::mapPgcTimeToPgcOverhead(MM_EnvironmentVLHGC *env, uintptr_t partialGcTimeMs) {
+	
+	/* Convert expectedTimeRatioMinimum/Maximum to 0-100 based for this formula */
+	double xminpct = _extensions->dnssExpectedTimeRatioMinimum._valueSpecified * 100;
+	double xmaxpct = _extensions->dnssExpectedTimeRatioMaximum._valueSpecified * 100;
+	double xmaxt = (double)_extensions->tarokTargetMaxPauseTime;
+
+	double overhead;
+
+	if (heapIsFullyExpanded(env)) {
+		/* 
+		 * Eden size is being driven by heuristic which is trying to MINIMIZE hybrid overead. A low avg pgc time, is more desirable than high avg time,
+		 * So the overhead logic needs to map a low avg pgc time, to a low overhead value (aka, a "better"/more desirable value)
+		 * Ex 20ms -> 5% (good/desirable), 1000ms -> 80% (bad/undesirable/eden should probably shrink)
+		 */
+		double midpointPct = (xmaxpct + xminpct)/2.0;
+		if (partialGcTimeMs <= xmaxt) {
+			/* Once the pgc time is at, or below the max pgc time, there is no "benefit" from shrinking it further, since we are already satisfying tarokTargetMaxPauseTime */
+			overhead = midpointPct;
+		} else {
+			/* 
+			 * If pgc time is above the max pgc time, map high PGC time values as very very high overhead, in efforts to bring the PGC time down to tarokTargetMaxPauseTime
+			 * If pgc time is only slightly above tarokTargetMaxPauseTime, then there is only a very small overhead penalty, 
+			 * wheras being 2x higher than the target pause time leads to a significantly bigger penalty  
+			 */
+			double overheadCurve = pow(1.03, ((double)partialGcTimeMs - xmaxt)) + midpointPct - 1;
+			overhead = OMR_MIN(100.0, overheadCurve);
+		}
+		
+	} else {
+		/* 
+		 * Eden sizing logic is trying to keep hybrid overhead between xminpct and xmaxpct, while trying to respect xmaxt. 
+		 * In this situation, when pgc times are very high (above xmaxt), the overhead score needs to return a low number, suggesting contraction.
+		 * If partialGcTimeMs is less than half of xmaxt, eden can expand without any fear of getting close to xmaxt - the mapped cpu overhead here is > xmaxpct (suggesting eden expansion)
+		 * Ex: 20ms -> 12% (suggest expansion), 2000ms -> 0.00% (suggest contraction)
+		 */
+		double slope = (xmaxpct - xminpct) / ((xmaxt/2) - xmaxt);
+		overhead = (slope * (double)partialGcTimeMs) + ((2.0 * xmaxpct) - xminpct);
+		overhead = OMR_MAX(overhead, 0.0);	
```

Comment:
if I calculated correctly
- with partialGcTimeMs >= xmaxt, this should return a value that is <= xminpct (hence explicitely suggesting contract)
- as partialGcTimeMs is approaching xmaxt/2, the return value is approaching xmaxpct (suggesting neither explicit contraction nor explicit expansion, although blended later with CPU overhead it may implicitly result into a contraction/expansion)
- with partialGcTimeMs  < xmaxt/2, the returned value is xmaxpct (still not explicitely suggesting expansion, but may quite help to occur)
It think it reasonable, although hard to quantify without trying some other mappings

If my comments are correct/sensible, you can consider embedding (and correcting/adjusting if needed) them into the code.

Reply:
Yes thats correct.

One thing I have been thinking about, is whether or not `overhead = OMR_MAX(overhead, 0.0);` is desirable or not. If `partialGcTimeMs ` is far over the pause time target (whether it be 2x higher, or 100x higher), it will map to 0. If we allow this mapping to go negative, it will likely get a better balanced of pgc time & cpu overhead - despite a negative overhead not really making sense (from a high level point of view). Thoughts on removing the line which keeps the mapped overhead to >0?

Reply:
As said earlier, I can't tell how aggressively this mapping will suggest contraction (or expansion) without seeing it in action (and comparing with some other mapping). 
Being a new thing, let's play conservative (be fairly soft) in regards to the pause criteria, and keep the cap (keep it non-negative). Down the road (when we get some real life experience) we may consider some tuning.